### PR TITLE
Remove request for preprocessor guards from header files.

### LINKED ISSUE #340: [BUG] Follow up on #321
Pull Request #321 was merged with a few nits to be addressed. Also, it did not include a changelog entry.

This is a blocker for ACLE release 2024Q3.

### PR DESCRIPTION
**Pull request by CarolineConcatto**
---
name: Pull request
about: Technical issues, document format problems, bugs in scripts or feature proposal.

---

**Thank you for submitting a pull request!**

If this PR is about a bugfix:

Please use the bugfix label and make sure to go through the checklist below.

If this PR is about a proposal:

We are looking forward to evaluate your proposal, and if possible to
make it part of the Arm C Language Extension (ACLE) specifications.

We would like to encourage you reading through the [contribution
guidelines](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md), in particular the section on [submitting
a proposal](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md#proposals-for-new-content).

Please use the proposal label.

As for any pull request, please make sure to go through the below
checklist.

Checklist: (mark with ``X`` those which apply)

* [ ] If an issue reporting the bug exists, I have mentioned it in the
      PR (do not bother creating the issue if all you want to do is
      fixing the bug yourself).
* [ ] I have added/updated the `SPDX-FileCopyrightText` lines on top
      of any file I have edited. Format is `SPDX-FileCopyrightText:
      Copyright {year} {entity or name} <{contact informations}>`
      (Please update existing copyright lines if applicable. You can
      specify year ranges with hyphen , as in `2017-2019`, and use
      commas to separate gaps, as in `2018-2020, 2022`).
* [ ] I have updated the `Copyright` section of the sources of the
      specification I have edited (this will show up in the text
      rendered in the PDF and other output format supported). The
      format is the same described in the previous item.
* [ ] I have run the CI scripts (if applicable, as they might be
      tricky to set up on non-*nix machines). The sequence can be
      found in the [contribution
      guidelines](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md#continuous-integration). Don't
      worry if you cannot run these scripts on your machine, your
      patch will be automatically checked in the Actions of the pull
      request.
* [ ] I have added an item that describes the changes I have
      introduced in this PR in the section **Changes for next
      release** of the section **Change Control**/**Document history**
      of the document. Create **Changes for next release** if it does
      not exist. Notice that changes that are not modifying the
      content and rendering of the specifications (both HTML and PDF)
      do not need to be listed.
* [ ] When modifying content and/or its rendering, I have checked the
      correctness of the result in the PDF output (please refer to the
      instructions on [how to build the PDFs
      locally](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md#continuous-integration)).
* [ ] The variable `draftversion` is set to `true` in the YAML header
      of the sources of the specifications I have modified.
* [ ] Please *DO NOT* add my GitHub profile to the list of contributors
      in the [README](https://github.com/ARM-software/acle/blob/main/README.md#contributors-) page of the project.

### CODE REVIEW

File: main/acle.md
Comment (paulwalker-arm) [inline main/acle.md:1026]: I believe this should remain because it's a test if the compiler supports this software extension rather than something related to a target feature.

File: main/acle.md
Comment (mgabka) [inline main/acle.md:1083]: this is badly phrased.  Do you mean here that users should not rely on #pragma or attribute or the feature.
macros?

If it is about the macros, I think we need to phrase it differently and explain that the presence of the macro does not guarantee that the architecture feature is always present at various stages of compilation or execution.

File: main/acle.md
Comment (mgabka) [inline main/acle.md:1093]: don't you need to extend the endif to cover the definition of foo?

File: main/acle.md
Comment (CarolineConcatto) [reply main/acle.md]: Hi Maciej,
I re-wrote this paragraph and added a section at the end  of HEADER to explain headers + macros. 
Just in case the ACLE has a section: **Feature test macros**. It is  generic, it is about all predefined macros. 
The point here is that  the ACLE/I  want to explain that  headers protected by macro  does not imply we can always use intrinsics.

File: main/acle.md
Comment (paulwalker-arm) [inline main/acle.md:1102]: How about?

Testing a feature macro returns the compiler's support for the feature via ACLE documented intrinsics and inline assembly. This property can be independent of where the test is performed, meaning a feature macro does not guarantee the validity of using the feature at the point it is tested. For example:

``` c
    #include <arm_sme.h>

    void foo(svbool_t pg, void *ptr, uint32_t slice_base) {
    #ifdef __ARM_FEATURE_SME
      svst1_hor_za8(0, slice_base, pg, ptr);
    #endif
    }
```

Whilst testing \`__ARM_FEATURE_SME\` ensures the compiler supports the SME intrinsic \`svst1_hor_za8\`, \`foo\` will fail to compile because it has not been decorated with the necessary keywords to guarantee the function will be executed in streaming mode.

Not all such issues can be caught during compilation and may instead result in runtime failures.

File: main/acle.md
Comment (mgabka) [inline main/acle.md:1077]: should this be changed to : "### predefined feature macros and header files" ?

File: main/acle.md
Comment (mgabka) [reply main/acle.md:1102]: > Testing a feature macro returns the compiler's support for the feature via ACLE documented intrinsics and inline assembly.

Not sure if it is just me, the key is how we define "support".
For me compiler supports ACLE when the given ACLE is implemented. While the feature macro is only defined when appropriate features are enabled in the cmd line option only. Maybe worth to change it to sth like:

"Testing a feature macro allows to check if the compiler can make use of the architecture feature via ACLE documented intrinsics and inline assembly assuming the provided compilation flags."

What do you think?

Comment (paulwalker-arm) [reply main/acle.md:1102]: This distinction is not relevant because there is no assumption about compilation flags.  The macro alone is sufficient to detect at the source level whether the compiler (in the mode it's in when compiling the input) can support a specific feature.  By support we effectively mean "is able to compile the input when it is correctly formed".

The provided example shows something that is not "correctly formed" where the compiler will report an error that must be fixed before the code will compile. There are many other examples but I don't see a reason to document them all here.

Note I say "compile" and not "compile to a working binary" because the more typical result is the compiler correctly building the input which then generates a SIGILL when executed. This is covered by the "Not all such issues..." sentence.  Personally I don't think it's worth going into anymore detail than this.

Comment (mgabka) [reply main/acle.md:1102]: Sure, I get the difference between "compile" and not "compile to a working binary". 

For me it was just the meaning of word support, as in my mind feature could be supported (as implemented) by a tool but not available when given set of options it provided. It has this small difference from the user point of view as if feature is not implemented they can not do anything, while if it is not enabled they can change options.
But I am not against your version either.

Comment (paulwalker-arm) [reply main/acle.md:1102]: I think that's because you are considering this from the point of view of a human asking whether the compiler can support a feature. Whereas the feature macros are source code actions and thus the context is the point the source code is being compiled, at which point the compiler has already been "configured" (for want of a better word) by the user.

Perhaps the point you are raising is the document does not detail what actions the user should take in order to control the definitions of the feature flags, which is intentional because we cannot dictate that.

Comment (rsandifo-arm) [reply main/acle.md:1102]: I agree it would be good to be a bit more precise about what “support” means though.  At one level, GCC 14 “supports” SME2, and the word is often used in that sense.  So if we just say “support”, users might expect GCC 14 to define `__ARM_FEATURE_SME2` unconditionally.

Instead, for compatibility, we have to define the macro only if the intrinsics are “available” in functions that have no target annotations (not precise wording either).  What users need to do to achieve that for a given compiler is outside the scope of the ACLE, but I think the result is within scope.

IMO, the example of using SME intrinsics in non-streaming functions is an example of an intrinsics being available, but being used incorrectly.  It's somewhat like passing an out-of-range value to a constant argument, except that the error is contextual, rather than contained within the call.  (I still think it's a good example though.)

**Comment by CarolineConcatto** [comment]
Thank you @paulwalker-arm, @mgabka  and @rsandifo-arm  for your inputs. 
I have made the changes asked.
Carol

File: main/acle.md
Comment (CarolineConcatto) [reply main/acle.md]: I did changed to what @paulwalker-arm requested, but instead of using support I wrote available, as suggested by @rsandifo-arm.
I hope it is correct.

**Review by DanielKristofKiss — APPROVED** [review]
LGTM

**Comment by vhscampos** [comment]
Let's wait until the end of this week for additional feedback

**Review by sdesmalen-arm — CHANGES_REQUESTED** [review]

File: main/acle.md
Comment (rsandifo-arm) [reply main/acle.md:930]: Agreed, but also: `__ARM_ACLE` is a special case in that, like `__ARM_NEON_SVE_BRIDGE`, it describes the presence of a header file rather than a specific architecture feature.  So I think the original text was ok too (and so is your suggestion).

For the feature macros, maybe “is available in a normal unannotated function” would be more specific than “is available”.

File: main/acle.md
Comment (rsandifo-arm) [inline main/acle.md:1091]:
```suggestion
Feature macros indicate whether that feature's intrinsics and inline assembly instructions
are available in normal unannotated functions. In some implementations, the macro definitions
are invariant throughout a translation unit, while in other implementations, the macro
definitions can vary with pragmas.

The macro definitions are otherwise independent of where the test is performed,
meaning that a feature macro does not guarantee that the feature can be used at the
point that it is tested. For example:
```

File: main/acle.md
Comment (rsandifo-arm) [inline main/acle.md:1106]:
```suggestion
While testing `__ARM_FEATURE_SME` ensures that the SME intrinsic `svst1_hor_za8`
is available, `foo` will fail to compile because the call does not occur in a
[streaming statement](#streaming-statement).
```

File: main/acle.md
Comment (rsandifo-arm) [inline main/acle.md:1109]: Do you know of a specific example where that's true?

File: main/acle.md
Comment (CarolineConcatto) [reply main/acle.md:930]: So this was suggested by @paulwalker-arm 

The `XYZ` macro can be tested before including the header. However, this is not recommended for cases where target feature selection is overridden within code, for example, when using #function-multi-versioning.
From another previous asked change.
 
I have no preference, but would be nice to agree in what to write here, before I change again.  The sentence you see here is a mix of previous interaction that we agree we should replace taking into account FMV.

File: main/acle.md
Comment (CarolineConcatto) [reply main/acle.md:1087]: The goal here is to show that guarding  code with macros does not guarantee  a working binary, because this intrinsics may need more than that.

File: main/acle.md
Comment (CarolineConcatto) [reply main/acle.md:1112]: I cannot think of anything that we have now.
The frontend is doing a decent job protecting the intrinsics at compiler time. 
I can remove if you think is best

File: main/acle.md
Comment (rsandifo-arm) [reply main/acle.md:930]: I'm not sure that wording is appropriate for `__ARM_ACLE` though (specifically that macro, not others).  There is no change in features that would affect whether the compiler provides `<arm_acle.h>`.  Intrinsics within `<arm_acle.h>` that depend on features have their own `__ARM_FEATURE_*` macro.

File: main/acle.md
Comment (rsandifo-arm) [reply main/acle.md:1112]: Yeah, think it would be best to remove it if we can't give a specific example.

File: main/acle.md
Comment (sdesmalen-arm) [reply main/acle.md:1087]: The current description is better, but I wonder if it would be sufficient just to say: "No assumptions should be made on the order or context in which the preprocessor macros are evaluated, for example:"

What do you think @CarolineConcatto @rsandifo-arm  ?

File: main/acle.md
Comment (sdesmalen-arm) [inline main/acle.md:1103]: I thought the paragraph above is more about the evaluation context/order of the macro to express that a user can't rely on the macro being set/unset when used with the ACLE keyword attributes, e.g.

```
__attribute__((target("+sve")))
void foo() {
#ifdef __ARM_FEATURE_SVE
  // The user should make no assumptions that the target attribute
  // has enabled the __ARM_FEATURE_SVE macro.
#endif
}
```

The current example is more about showing that the compiler can add additional restrictions to the use of an intrinsic beyond what is captured by the ACLE macro. If that is the intent of the example, I think it would be helpful if that can be added as an separate paragraph, e.g. something like "The compiler may add additional restrictions to the intrinsics beyond what is captured by the ACLE macros depending on the context in which the intrinsics are used".

Comment (CarolineConcatto) [reply main/acle.md]: Thank you @sdesmalen-arm  for the suggestions.
I made some changes according to your comments.

File: main/acle.md
Comment (sdesmalen-arm) [inline main/acle.md:943]: The macro is also available outside of function scope. The scope in which it is used shouldn't matter, because it is a preprocessor macro.

File: main/acle.md
Comment (sdesmalen-arm) [inline main/acle.md:1082]: I think it's worth including the other example I shared here.

File: main/acle.md
Comment (sdesmalen-arm) [inline main/acle.md:1081]: I find the word 'test' confusing, would it make more sense to speak of 'evaluating' ?

```suggestion
Evaluating a feature macro returns the availability of intrinsics and inline
assembly for that feature, but no assumptions should be made on the order or
context in which the preprocessor macros are evaluated.
```

File: main/acle.md
Comment (sdesmalen-arm) [inline main/acle.md:1098]:
```suggestion
If `__ARM_FEATURE_SME` evaluates to `true` the SME intrinsic `svst1_hor_za8`
is available, but `foo` may still fail to compile because the call does not occur in a
[streaming statement](#streaming-statement).
```

File: main/acle.md
Comment (CarolineConcatto) [reply main/acle.md]: I wrote again this sentence to:
When `__ARM_xxx` is defined to `1`, the header file is available
regardless of the context in which the macro is evaluated.

File: main/acle.md
Comment (sdesmalen-arm) [inline main/acle.md:1026]: nit: unnecessary change.

File: main/acle.md
Comment (sdesmalen-arm) [inline main/acle.md:1090]: nit: whitespace
```suggestion
      // The user should make no assumptions that the target attribute
      // has enabled the __ARM_FEATURE_SVE macro.
```

File: main/acle.md
Comment (sdesmalen-arm) [inline main/acle.md:1097]: I'm still not convinced this is worth pointing out, because the compiler may also enforce requirements on the range of immediate arguments, and we never bothered to document that either. That said though, I don't see any harm in adding this either.

**Review by sdesmalen-arm — APPROVED** [review]
I'm content with the phrasing now. Please address the nits though.

**Comment by vhscampos** [comment]
Since this PR got approval with minor nits, I'm merging it. @CarolineConcatto please address the remaining nits in a following PR. Thanks

File: main/acle.md
Comment (sallyarmneale) [inline main/acle.md:1109]: may -> might
https://confluence.arm.com/display/TCOM/Concise+and+clear+words+and+phrases

**Review by sallyarmneale — COMMENTED** [review]
Two minor changes

**Pull request merged**
